libraries/AC_Fence/AC_PolyFence_loader.cpp

```diff
@@ -106,27 +229,1307 @@ bool AC_PolyFence_loader::breached(const Location& loc)
+        gcs().send_text(MAV_SEVERITY_WARNING, "Fence exceeds storage size");
+        return false;
+    }
+
+    if (!format()) {
+        return false;
+    }
+
+    uint8_t total_vertex_count = 0;
+    uint16_t offset = 4; // skipping magic
+    uint8_t vertex_count = 0;
+    for (uint16_t i=0; i<count; i++) {
+        const AC_PolyFenceItem new_item = new_items[i];
+        switch (new_item.type) {
+        case AC_PolyFenceType::POLYGON_INCLUSION:
+        case AC_PolyFenceType::POLYGON_EXCLUSION:
+            if (vertex_count == 0) {
+                // write out new polygon count
+                vertex_count = new_item.vertex_count;
+                total_vertex_count += vertex_count;
+                if (!write_type_to_storage(offset, new_item.type)) {
+                    return false;
+                }
+                fence_storage.write_uint8(offset, vertex_count);
+                offset++;
+            }
+            vertex_count--;
+            if (!write_latlon_to_storage(offset, new_item.loc)) {
+                return false;
+            }
+            break;
+        case AC_PolyFenceType::END_OF_STORAGE:
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+            AP_HAL::panic("asked to store end-of-storage marker");
+#endif
+            return false;
+        case AC_PolyFenceType::CIRCLE_INCLUSION:
+        case AC_PolyFenceType::CIRCLE_EXCLUSION:
+            total_vertex_count++; // useful to make number of lines in QGC file match FENCE_TOTAL
+            if (!write_type_to_storage(offset, new_item.type)) {
+                return false;
+            }
+            if (!write_latlon_to_storage(offset, new_item.loc)) {
+                return false;
+            }
+            // store the radius
+            fence_storage.write_uint32(offset, new_item.radius);
+            offset += 4;
+            break;
+        case AC_PolyFenceType::RETURN_POINT:
+            if (!write_type_to_storage(offset, new_item.type)) {
+                return false;
+            }
+            if (!write_latlon_to_storage(offset, new_item.loc)) {
+                return false;
+            }
+            break;
+        }
+    }
+    if (!write_eos_to_storage(offset)) {
+        return false;
+    }
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    if (!index_eeprom()) {
+        AP_HAL::panic("Failed to index eeprom");
+    }
+    gcs().send_text(MAV_SEVERITY_DEBUG, "Fence Indexed OK");
+#endif
+
+    void_index();
+
+    // this may be completely bogus total.  If we are storing an
+    // advanced fence then the old protocol which relies on this value
+    // will error off if the GCS tries to fetch points.  This number
+    // should be correct for a "compatible" fence, however.
+    uint16_t new_total = 0;
+    if (total_vertex_count == 0) {
+        new_total = 0;
+    } else if (total_vertex_count < 3) {
+        // we fake things up in FENCE_POINT fetch and send
+        new_total = 5;
```

Comment:
Excepting 0, 5 is the minimum valid value for the old fence protocol (1 return point, 3 for a polygon, 1 for the closing point).

Since we can't bounce the invalid set, this is our best option - fix the parameter after the fact.

Note that this should be a can't-happen - the fence validation code should have picked up any short polygon.  Now it's basically defense-in-depth.

Reply:
I've removed this compatibility code.  It's a can't-happen anyway, as previously noted.

If there are less than three points we squash it to zero.